service/controller/v14/resource/bridgezone/resource.go

```diff
@@ -0,0 +1,415 @@
+//	    ├── A api.new_cluster.k8s.installation.eu-central-1.aws.gigantic.io
+//	    └── A ingress.new_cluster.k8s.installation.eu-central-1.aws.gigantic.io
+//
+// NOTE: In the code below k8s.eu-central-1.aws.gigantic.io zone is called
+// "intermediate" and new_cluster.k8s.eu-central-1.aws.gigantic.io zone is
+// called "final". This resource *only* ensures we have delegation from the
+// intermediate zone to the final zone, but only if the intermediate zone
+// exists.
+//
+// After we have guest clusters managed by this resource set (v14) and newer
+// the DNS layout should look like:
+//
+//	installation.eu-central-1.aws.gigantic.io (host account)
+//	├── NS k8s.installation.eu-central-1.aws.gigantic.io (default guest account)
+//	│   └── NS new_cluster.k8s.installation.eu-central-1.aws.gigantic.io (byoc guest account)
+//	└── NS new_cluster.k8s.installation.eu-central-1.aws.gigantic.io (byoc guest account)
+//	    ├── A api.new_cluster.k8s.installation.eu-central-1.aws.gigantic.io
+//	    └── A ingress.new_cluster.k8s.installation.eu-central-1.aws.gigantic.io
+//
+// At this point we should be fine with removing
+// k8s.installation.eu-central-1.aws.gigantic.io NS record from
+// installation.eu-central-1.aws.gigantic.io zone. Then after a couple of days
+// when delegation propagates and DNS caches are refreshed we can delete
+// k8s.installation.eu-central-1.aws.gigantic.io zone from the default guest
+// account.
+type Resource struct {
+	hostAWSConfig aws.Config
+	hostRoute53   *route53.Route53
+	k8sClient     kubernetes.Interface
+	logger        micrologger.Logger
+
+	route53Enabled bool
+}
+
+func New(config Config) (*Resource, error) {
+	if reflect.DeepEqual(aws.Config{}, config.HostAWSConfig) {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostAWSConfig must not be empty", config)
+	}
+	if config.HostRoute53 == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostRoute53 must not be empty", config)
+	}
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		hostAWSConfig: config.HostAWSConfig,
+		hostRoute53:   config.HostRoute53,
+		k8sClient:     config.K8sClient,
+		logger:        config.Logger,
+
+		route53Enabled: config.Route53Enabled,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return name
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !r.route53Enabled {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "route53 disabled, skipping execution")
+		return nil
+	}
+
+	baseDomain := key.BaseDomain(customObject)
+	intermediateZone := "k8s." + baseDomain
```

Comment:
I feel like the domain construction is now just done everywhere. Maintaining this is no fun. How can we make this better?